### PR TITLE
[failproofai-535] fix(docs): convert HTML comments to MDX comments in translated READMEs

### PR DESCRIPTION
**Pull request by NiveditJain**
## Problem

The **Mintlify Deployment** check on `main` (commit `446484b`) [failed](https://github.com/FailproofAI/failproofai/runs/87589878138): every one of the 14 `docs/i18n/README.*.md` translations failed to publish with

> Failed to parse page content at path i18n/README.ar.md: Unexpected character `` ! `` (U+0021) before name, expected a character that can start a name … (note: to create a comment in MDX, use `{/* text */}`)

## Root cause

Mintlify parses every page as **MDX**, and the README carries a top-level HTML comment:

```
<!-- A 6-column table instead of inline <img> runs: table columns never re-wrap, ... -->
```

HTML comments (`<!-- ... -->`) are a hard MDX syntax error. The comment lives in the root `README.md` (fine — GitHub renders it invisibly) and the README translator copies it verbatim into each `docs/i18n/README.*.md`, which Mintlify then rejects.

## Fix

Add a fence-aware `convertHtmlComments` sanitizer — a sibling of the existing `stripStrayTrailingFence` / `sanitizeJsxAttributes` MDX sanitizers — that rewrites `<!-- ... -->` → `{/* ... */}` on translator output:

- Wired into both `readme-translator.ts` and `translateMdxPage`, so future translation runs stay green.
- **Fence-aware**: `<!-- -->` *inside* code fences is left literal (e.g. the `` ```xml `` plist snippet in the AgentEye collector docs), so those pages are untouched.
- Neutralizes any `*/` inside a comment body so the generated MDX comment can't terminate early.
- The 14 already-committed translations are regenerated to match (only the comment delimiters change).

The root `README.md` keeps HTML-comment syntax (GitHub-facing); only the Mintlify-served copies under `docs/i18n/` are rewritten.

## Hardening — why CI didn't catch this

The `validate:mdx` CI job (`scripts/validate-mdx.ts`) exists precisely to parse every docs page with Mintlify's engine on the PR, before the post-merge deploy. But it only walked **`.mdx`** files — and the i18n READMEs are **`.md`** (which Mintlify also parses as MDX), so their breakage sailed straight past CI to the deploy. This PR:

- Extends `collectMdxFiles` to validate `.md` pages too, so this whole class of translation breakage fails on the PR instead of `main`. (Note: Mintlify runs no preview deploy on PRs, so this is the only pre-merge validation of the i18n pages.)
- Aligns `readme-translator.ts` with `translateMdxPage` by also running `sanitizeJsxAttributes` over its JSX-bearing output (the logo table).

## Testing

- Unit tests for `convertHtmlComments` (single/multi-line, `` in body, fenced-vs-top-level, `*/` neutralization, no-op).
- `collectMdxFiles` test (`.md` + `.mdx` collected, other files ignored).
- `findMdxParseError` regressions: a top-level `<!-- -->` fails to parse; the `{/* */}` form and a fenced `<!-- -->` parse cleanly (same engine Mintlify uses).
- Verified `grep -rn '<!--' docs/i18n/` returns **0** after regeneration; the i18n diff is limited to the two delimiter lines per file.
- Full lint / tsc / test / e2e via CI.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Eg5Bgx3ucKNJNZ2f2Ng5dQ

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed localized README translation output by converting top-level HTML comments to MDX-compatible comment syntax to prevent documentation deployment failures.
  - Preserved comment-like content inside fenced code blocks and regenerated the affected localized READMEs.
  - Expanded CI MDX validation so both `.mdx` and `.md` pages are checked as MDX.

- **Tests**
  - Added coverage for single-line and multi-line comment conversion, safe handling of nested/edge cases, code-fence preservation, and MDX parse-error detection; included regression coverage for collecting `.md`/`.mdx` files.

### CODE REVIEW

**Comment by hermes-exosphere** [comment]
Automated code review started - full review. Results will be posted here.

**Comment by hermes-exosphere** [comment]
## Automated Code Review

### Executive Summary
This PR fixes a Mintlify docs deployment failure where all 14 `docs/i18n/README.*.md` translations failed to publish due to HTML comments (`<!-- ... -->`) being a hard MDX syntax error. A new fence-aware `convertHtmlComments` sanitizer rewrites HTML comments to MDX comments (`{/* ... */}`) in translator output, with proper handling of fenced code blocks and `*/` neutralization. The fix is surgical, well-tested, and non-breaking — clean code with no issues found.

---

### Change Architecture

```mermaid
graph TD
    A["README.md (root)"] -->|"readme-translator.ts"| B["translateReadme()"]
    C["English MDX pages"] -->|"mdx-translator.ts"| D["translateMdxPage()"]
    B --> E["convertHtmlComments()"]
    D --> E
    E --> F["<!-- --> becomes {/* */}"]
    E --> G["Fence-aware: skip inside fences"]
    E --> H["Neutralize */ to * /"]
    F --> I["14 i18n READMEs regenerated"]
    style E fill:#90EE90
    style I fill:#90EE90
    style A fill:#87CEEB
    style C fill:#87CEEB
```

**Legend:** Green = New | Blue = Modified

---

### Breaking Changes
No breaking changes detected. All changes are additive — new sanitizer, no API or schema modifications.

---

### Issues Found
No issues found. Code is clean and well-structured.

---

### Logical / Bug Analysis

**`convertHtmlComments` function** (`scripts/translate-docs/mdx-translator.ts:112-141`):
- Fence-aware detection: uses even/odd marker count to detect if a comment sits inside a code block. Works correctly for non-nested fences (the README's only use case).
- `*/` neutralization: body.replace(`/\*\//g`, `* /`) prevents premature MDX comment closure.
- Lazy regex `<!--([\s\S]*?)-->/g` correctly captures each comment individually, including multi-line and multiple comments on the same line.
- Returns identity for input with no comments (no-op).

**Known limitation** (non-blocking): Nested fences (e.g., outer 4-tick block containing inner 3-tick block) would miscount the fence depth since the function uses flat even/odd rather than a stack. This does not affect the README (zero nested fences) and is documented implicitly by the code structure.

**Integration points**:
- `readme-translator.ts:110`: `convertHtmlComments(stripStrayTrailingFence(translated))` — applied after other sanitizers, before output assembly.
- `mdx-translator.ts:226`: Same pipeline applied in `translateMdxPage()`.

**14 regenerated translations** — all changes are consistent: only two lines per file change (opening/closing comment delimiters). Verified with `grep -rn '<!--' docs/i18n/` returning 0 matches.

---

### Evidence — Build & Test Results

<details>
<summary>Test Results</summary>

```
RUN  v4.1.10

Test Files  13 passed (13)
     Tests  147 passed (147)
  Duration  4.66s
```

</details>

<details>
<summary>Syntax Check</summary>

```
node --check scripts/translate-docs/mdx-translator.ts  -> OK
node --check scripts/translate-docs/readme-translator.ts -> OK
```

</details>

<details>
<summary>HTML Comment Verification</summary>

```
grep -rn '<!--' docs/i18n/
(no matches — all 14 files clean)
```

</details>

**New tests for `convertHtmlComments`** — 7 test cases added:
- Single-line comment conversion
- Multi-line comment with preserved content
- `` inside comment body (regression guard)
- Fenced code block preservation
- Top-level + nested fence mixing
- `*/` neutralization
- No-op for comment-free content

---

### Issue Linkage
No linked issue. This PR fixes a CI deployment failure (Mintlify). Consider linking to the failed CI run for full traceability.

---

### Human Review Feedback
No human review comments on this PR.

---

### Suggestions
1. **Consider stack-based fence tracking** — the current flat even/odd approach works for all real use cases but would silently convert comments inside nested fences. If the repo ever adds nested code blocks to docs, this could surface. A minimal stack tracking the fence depth (pushing on open, popping on matching close) would make the function fully general. Not blocking — the README has zero nested fences.
2. **Edge case: `<!--` inside comment body** — the lazy regex `<!--([\s\S]*?)-->/g` stops at the first `-->`. If a comment body contains the literal text `<!--`, the inner `<!--` would be treated as the start of a new comment, leaving the preceding text orphaned. This is purely theoretical for the current README content.
3. **Test the `readme-translator.ts` integration** — the readme-translator tests currently only cover `buildMainReadmeLanguageLinks`. Consider adding a test that verifies `translateReadme()` calls `convertHtmlComments` on the translated output (e.g., mock `translateContent` to return a string with HTML comments and assert MDX output).

---

### Verdict

VERDICT: APPROVED

---

*Automated code review*

**Review by hermes-exosphere — APPROVED** [review]
Automated review: Approved. All 147 tests pass, no issues found. Clean implementation of MDX comment sanitization.

**Review by hermes-exosphere — DISMISSED** [review]
Automated review: Approved. ✅

**Comment by hermes-exosphere** [comment]
Automated code review started - full review. Results will be posted here.

File: scripts/translate-docs/mdx-translator.ts [hunk 1]
```diff
@@ -94,6 +94,52 @@ export function stripStrayTrailingFence(content: string): string {
...
+  return content.replace(
+    /<!--([\s\S]*?)-->/g,
+    (match: string, body: string, offset: number) => {
+      if (isInsideFence(offset)) return match;
+      // Neutralize any `*/` so it can't close the JS block comment early.
+      const safeBody = body.replace(/\*\//g, "* /");
+      return `{/*${safeBody}*/}`;
+    },
+  );
+}
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply scripts/translate-docs/mdx-translator.ts:149]: Fixed in 1d72e4b. Replaced the generic fence-toggle counter with CommonMark-style tracking: a fence now closes only on a later line using the same fence character and a length >= the opener; a different character or shorter marker while a fence is open is treated as inner content, and an unterminated fence runs to EOF. This correctly handles a 4-backtick block embedding a 3-backtick block and mixed backtick/tilde fences, so a real top-level comment after such a block is still converted. Added regression tests for all three cases.

**Comment by hermes-exosphere** [comment]
⏳ **Automated re-review in progress...** Deep analysis of all 3 commits (1d72e4b, 7ae1d99, 71a6352) across 20 files. Running tests and verifying.

**Comment by hermes-exosphere** [comment]
## 🔍 Automated Re-Review

### 📋 Executive Summary
This is a re-review of PR #535 which fixes a Mintlify docs deployment failure by converting HTML comments to MDX comments in translated READMEs. The previous review (hermes-exosphere) approved with suggestions. Since then, two additional commits addressed the CodeRabbit-identified fence-tracking issue (1d72e4b replaced toggle counting with CommonMark-style character+length tracking) and extended CI validation (7ae1d99). The code is clean, well-tested, and all 2078 tests pass. No regressions.

---

### 🔴 Breaking Changes
No breaking changes detected. All changes are additive — new function, extended file glob, modified sanitizer pipelines applied to translator output. No API, schema, config, or dependency changes.

---

### ⚠️ Issues Found
No critical or warning-level issues found. Three minor suggestions below.

---

### 🔬 Logical / Bug Analysis

**convertHtmlComments function** (scripts/translate-docs/mdx-translator.ts:112-149):
- Fence tracking (lines 119-135): Correctly tracks active fences by character and opening-marker length per CommonMark. A closing fence requires same character + length >= opener. Mixed ~~~ inside ``` blocks are correctly treated as inner content. Unterminated fences extend to EOF. This was the CodeRabbit-identified issue, now fixed (commit 1d72e4b).
- Fence range end (line 128): Uses content.indexOf("\n", fenceRe.lastIndex) — correctly captures to end of the closing-fence line.
- Comment conversion (lines 140-148): Lazy regex captures each HTML comment individually. */ neutralization prevents early MDX comment termination. No-op for comment-free content.

**readme-translator.ts pipeline (lines 110-117):**
- Now runs the same three sanitizers as translateMdxPage: sanitizeJsxAttributes -> stripStrayTrailingFence -> convertHtmlComments. This alignment is correct — the README emits JSX (logo table) and must pass Mintlify MDX parsing.

**validate-mdx.ts — collectMdxFiles (lines 102-110):**
- Now collects .md files alongside .mdx. This closes the gap where .md pages (i18n READMEs) were invisible to CI validation. The change is surgical (one || addition) and the export enables testing.

**14 i18n files:** Only the comment delimiters changed ( to */}). Verified grep -rn '<!--' docs/i18n/ returns zero matches — all HTML comments converted.

**CHANGELOG:** Two well-documented entries under "Fixes" — clear, comprehensive descriptions.

---

### 🧪 Evidence — Build & Test Results

Test Results (bun run test:run):
  121 test files passed (121)
  2078 tests passed (2078)
  Duration 72.63s
  Including: convertHtmlComments (10 new tests — all pass), collectMdxFiles (1 new test — passes), findMdxParseError (3 new tests for HTML comment detection — all pass)

TypeScript Check (tsc --noEmit): Clean — zero errors

Lint (eslint): 0 errors, 5 pre-existing warnings (none from changed files)

HTML Comment Verification: grep -rn '<!--' docs/i18n/ — no matches, all 14 files clean

---

### 🔗 Issue Linkage
No linked issue. This PR fixes a CI deployment failure (Mintlify). Consider linking to the failed CI run for full traceability.

---

### 👥 Human Review Feedback
- @NiveditJain (on thread PRRT_kwDOR6tUec6RdM0P): Confirmed fix in commit 1d72e4b — replaced generic fence-toggle counter with CommonMark-style tracking. This is addressed and verified in the current code.
- Previous hermes-exosphere review (id=4992528074): APPROVED with 3 suggestions:
  - Suggestion 1 (stack-based fence tracking): Addressed in 1d72e4b — now uses character+length tracking
  - Suggestion 2 (nested <!-- edge case): Still theoretical, not addressed (not needed for current README content)
  - Suggestion 3 (readme-translator integration test): Still a suggestion — existing tests pass

Both CodeRabbit review threads are already resolved (isResolved: true). No unresolved threads remain.

---

### 💡 Suggestions

1. Consider extracting fenceRe.lastIndex before content.indexOf (line 128, mdx-translator.ts). Currently it reads content.indexOf("\n", fenceRe.lastIndex) — fenceRe.lastIndex is stateful from the preceding exec() call. If anyone refactors this loop, forgetting to update this line would cause subtle bugs. A local variable would make the dependency explicit. Very minor — not blocking.

2. getEnglishMdxPages() still only collects .mdx (line 283: entry.endsWith(".mdx")). This is intentional (it feeds the translator, not the validator), but the name+behavior divergence from collectMdxFiles could confuse future maintainers. Consider a comment noting the distinction.

3. Integration test for readme-translator pipeline: The current tests verify buildMainReadmeLanguageLinks but not that translateReadme() actually calls convertHtmlComments + sanitizeJsxAttributes on translated output. A mock-based test would catch future regressions in the sanitizer pipeline.

---

### Verdict

VERDICT: APPROVED

---

Automated re-review · 2026-07-16 UTC

**Review by hermes-exosphere — APPROVED** [review]
Automated re-review: Approved. All 2078 tests pass, TypeScript compiles clean, lint clean, zero HTML comments remain in i18n docs. Both CodeRabbit threads are resolved. Three minor suggestions (non-blocking) in the summary comment above.

**Review by hermes-exosphere — APPROVED** [review]
Automated review: Approved. ✅
